### PR TITLE
[proof of concept] lazy load submodules for faster import time

### PR DESCRIPTION
**Pull request by smackesey**
This is a proof of concept for speeding Dagster's import time using lazy-loading of selected exports. Issue initially raised [here](https://threads.com/34418050050).

This PR implements lazy-loading for just `dagster.core.storage.event_log`. This was chosen bc it accounts for ~25% of Dagster's import time. You can confirm this like so (note `-X importtime` requires py 3.7+):

```
$ python -X importtime -c 'import dagster' # 800,715 usec (times will vary across machines)
$ DAGSTER_NO_LAZY_LOAD=1 python -X importtime -c 'import dagster'  # 1,103,590 usec
```

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6688** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/6688" target="_blank">Graphite</a> 
    * **PR #6719** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/6719" target="_blank">Graphite</a>  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

**Comment by alangenfeld** [comment]
Great work.

If we are using this pattern for deprecations i see no reason we shouldn't do this as well. 

Are there a handful more imports you can move to drop the cost even further?

File: python_modules/dagster/dagster/__init__.py [hunk 1]
```diff
@@ -253,19 +249,44 @@
...
-# that the typing.Annotated functionality introduced in Python 3.9 (or
-# available through the `typing-extensions` package) would allow us to avoid
-# the double declaration, e.g.:
-#
-# EventMetadataEntry: Annotated[Type[MetadataEntry], MetadataEntry, "0.15.0"]
+EventLogEntry: Any
+EventLogRecord: Any
+EventRecordsFilter: Any
+RunShardedEventsCursor: Any
+
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/__init__.py:269]: I guess this dimension is quite unfortunate

Comment (smackesey) [reply python_modules/dagster/dagster/__init__.py:269]: One more potential benefit of bringing in `typing-extensions` as dependency... (see longish comment about dynamic imports)

Comment (alangenfeld) [reply python_modules/dagster/dagster/__init__.py:269]: To clarify Im speaking to the type information being set to `Any` for these objects. I don't see how whats discussed in the comment would work for lazy loading. Am I missing something?

**Review by alangenfeld — COMMENTED** [review]
An aside from this proof of concept - these specific event classes could get moved around to get the rest of `core.storage.event_log` out of init import path

File: python_modules/dagster/dagster/__init__.py (hunk 1 above)
Comment (smackesey) [reply python_modules/dagster/dagster/__init__.py]: I think you're right on second thought-- but luckily it's now a moot point bc found a better solution using `typing.TYPE_CHECKING`.

**Comment by smackesey** [comment]
>What's the speedup on defining a simple op and directly invoking it?

**TL;DR: Simple top-level lazy loading does not improve load times for simple op execution at all, bc the various internal imports triggered when you call an op end up pulling in virtually all of Dagster. It is possible that extending lazy-loading to submodules could prevent this and lead to a significant speedup. Takes significant time to investigate so we should table this unless import time becomes a high priority issue.**

I compared the lazy/no-lazy import time difference when running this script:

```python
from dagster import op

@op
def hello_world():
    print('Hello World')

hello_world()
```

You can get the total dagster import time with this Unix command (it's complicated because when you don't just `import dagster`, you have to filter the multiple top-level dagster submodule imports and sum them):

```
python -X importtime simple_op.py 2>&1 | grep '| dagster' | cut -d '|' -f2 | paste -sd+ - | bc
```

Unfortunately, for the above script there is no apparent import time difference for lazy/no-lazy. From glancing over the output of `python -X importtime simple_op.py`, this is because calling an op pulls in basically all of Dagster. If you just _define_ the op without calling it (comment out last line in above script), then there's about a 2x speedup.

I was curious what would happen if we extended lazy load a little deeper, so I tried making `dagster.core.definitions` lazy load as well. This increased the speed-up from defining the op from 2x to 10x. 

I don't know whether calling the op actually _needs_ to pull in everything, I just know that it does. The fact that extending lazy-loading to `dagster.core.definitions` caused a significant speedup for op definition suggests that maybe there is some other module, which, if lazy loaded, would show a similar gain for op an op call. I don't have a good enough understanding of the whole system to know if that's the case.

Final note, because the import order changes when using lazy-loading, a few circular import issues surfaced that I had to track down and fix to get the script to execute. Any attempt to further explore would probably run into a bunch more of these.

**Comment by smackesey** [comment]
Closing for now since there is no clear path forward here.
